get reordering of columns on reference

ColeMiller1 opened the pull request:
Closes #4089. 

Current behavior will re-order columns when using ```(m)get``` even when ```.SDcols``` is present. This can be dangerous when updating in place:
```
library(data.table)
dt <- data.table(x = 1, y = 2, z = 3)
cols <- c("y","x")

dt[, (cols) := lapply(.SD[get("x") == 1],function(x){x}), .SDcols = cols ,by = z][]
       x     y     z
   <num> <num> <num>
1:     2     1     3

```
Proposed will not re-order the columns when ```.SDcols``` argument is present. 
```
dt[, (cols) := lapply(.SD[get("x") == 1],function(x){x}), .SDcols = cols ,by = z][]
       x     y     z
   <num> <num> <num>
1:     1     2     3
```
Also, some code was re-arranged to address a ```TODO``` comment related to this section of ```[.data.table```.

jangorecki commented on R/data.table.R:
```diff
@@ -968,7 +968,39 @@ replace_dot_alias = function(e) {
+            non_sdvars = setdiff(ansvars, sdvars)
+            ansvals = chmatch(ansvars, names_x)
+            if (verbose) 
+              cat("New:",paste(ansvars,collapse=","),"\n")
```
I would just add here `New:` -> `New ansvars:`, and similarly above in `Old:`

jangorecki commented on R/data.table.R:
```diff
@@ -968,7 +968,39 @@ replace_dot_alias = function(e) {
+          } else { ## TODO remove this else statement in Jan. 2023
+            newvars = setdiff(allcols, bynames) # checking if old cols c("y", "x") became c("x", "y")
+            if (length(newvars) == length(ansvars) && !all(newvars == ansvars)) {
+              warning("(m)get found in j with update assignment (i.e., `(cols) := ...`). j column order was re-ordered prior to version 1.12.9. See Issue #4089")
```
best to refer to even version numbers in messages

MichaelChirico replied:
Do we need to warn here? Since we're not changing any tests it could be considered as relying on untested behavior.

If we're going to warn here, there should be a way for users to turn it off. Properly I guess we'd have `options(datatable.warn.mget.reorder = TRUE)` for the next release, then change it `FALSE` the following release, remove the option the release after that?

MichaelChirico replied:
On re-read, this is what Jan was talking about in the comment on the main post:

https://github.com/Rdatatable/data.table/pull/4212#pullrequestreview-350862722

ColeMiller1 replied:
I removed the warning.

MichaelChirico commented on R/data.table.R:
```diff
@@ -968,7 +968,39 @@ replace_dot_alias = function(e) {
+        if (any(c("get", "mget") %chin% av)){
+          # get('varname') is too difficult to detect which columns are used in general
+          # eval(macro) column names are detected via the  if jsub[[1]]==eval switch earlier above.
+          b_warn = FALSE 
```
nitpick: i'm not understanding this variable name, what's it mean?

ColeMiller1 replied:
From my VBA days, b = boolean. Some habits are hard to get rid of. See also [Hungarian Notation](https://en.wikipedia.org/wiki/Hungarian_notation). 

Regarding the variable, I removed it.

MichaelChirico commented on R/data.table.R:
```diff
@@ -968,7 +968,39 @@ replace_dot_alias = function(e) {
+          }
+          if (!b_warn) {
+            if (verbose) 
+              cat("'(m)get' found in j. ansvars being set to all columns. Use .SDcols or a single j=eval(macro) instead. Both will detect the columns used which is important for efficiency.\nOld:", paste(ansvars,collapse=","),"\n")
```
Please convert this to use `cat(gettextf(..., domain="R-data.table"))` so it's caught for message translation later

MichaelChirico replied:
You might want to use `brackify(ansvars)` as well

MichaelChirico commented on R/data.table.R:
```diff
@@ -968,7 +968,39 @@ replace_dot_alias = function(e) {
+              cat("New:",paste(ansvars,collapse=","),"\n")
+          } else { ## TODO remove this else statement in Jan. 2023
+            newvars = setdiff(allcols, bynames) # checking if old cols c("y", "x") became c("x", "y")
+            if (length(newvars) == length(ansvars) && !all(newvars == ansvars)) {
```
the approach seems ex-post -- we changed the column order ourselves, right? where did that happen, and should we make the fix there instead?

ColeMiller1 replied:
This is where the change in column order happened.  the ```setdiff(allcols, bynames)``` re-orders the RHS. I decided to change the behavior here when ```!missing(.SDcols)``` - in these cases we want to prevent re-ordering.

ColeMiller1 commented:
Thank you both for the great comments. Main changes

1. Removed the warning
2. Updated to ```cat(gettextf("New ansvar ...", brackify(ansvars), domain = "R-data.table") 
3. Re-organized the comments and ```cat``` statements to better match the order of what I deleted.

mattdowle commented:
Quite an involved first time contribution! That part of the code is hard work, thank you. I've invited you to be project member; please accept invitation that shows in your github profile or projects tab. I'll add you to contributor list in a follow up commit.
